### PR TITLE
8296896: Change virtual Thread.yield to use external submit

### PR DESCRIPTION
**Pull request by AlanBateman**
The implementation of Thread.yield for virtual threads is currently a "lazy submit". This means the task for the virtual thread is queued to the carrier/worker local queue without signalling other threads. This behavior can be surprising/unfair when there are tasks in the submission queues, say when a platform thread has started or unparked a virtual thread.

Ron, Doug Lea, Viktor Klang and I have discussed this topic and propose to change Thread.yield to  use "external submit" when the local task queue is empty, and to push to the local queue when not empty. The change improves the fairness but will of course increase the chances that repeated Thread.yield will bounce between carriers.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296896](https://bugs.openjdk.org/browse/JDK-8296896): Change virtual Thread.yield to use external submit

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Ron Pressler](https://openjdk.org/census#rpressler) (@pron - Committer) ⚠️ Review applies to [2e707757](https://git.openjdk.org/jdk/pull/11533/files/2e707757f2a67769f38476a2c05ed5ef5c7da645)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11533/head:pull/11533` \
`$ git checkout pull/11533`

Update a local copy of the PR: \
`$ git checkout pull/11533` \
`$ git pull https://git.openjdk.org/jdk pull/11533/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11533`

View PR using the GUI difftool: \
`$ git pr show -t 11533`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11533.diff">https://git.openjdk.org/jdk/pull/11533.diff</a>

</details>

### CODE REVIEW

**Review by jaikiran — APPROVED** [review]
I'm not too familiar with the local queue and submission queues of the ForkJoinPool, but the PR description and the inline comments in the code/tests helped understand this change. With my limited knowledge of this area, this change looks OK to me.

I'm guessing the change to the existing `YieldALot` test to reduce the number of iterations from `500000` to `350000` is intentional to bring down the duration of that test?

**Review by pron — APPROVED** [review]

File: test/jdk/java/lang/Thread/virtual/YieldQueuing.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class YieldQueuing {
+
+    /**
+     * Test Thread.yield submits the task for the current virtyal thread to a scheduler
```
Comment (PaulSandoz) [inline test/jdk/java/lang/Thread/virtual/YieldQueuing.java:43]:
```suggestion
     * Test Thread.yield submits the task for the current virtual thread to a scheduler
```

Comment (AlanBateman) [reply test/jdk/java/lang/Thread/virtual/YieldQueuing.java]: Thanks, it's fixed now.

**Review by jaikiran — APPROVED** [review]

**Pull request closed without merging**